### DISCUSSION TITLE
does Optuna run the whole objective function for each trial?

### DISCUSSION THREAD

**Discussion by kaoutaar** [optuna/optuna#4588] [Q&A]
how does Optuna execute the objective function when calling `study.optimize(objective, n_trials=10`, 
is it, for each trial ===> call the function ??, if this is the case, this means we shouldn't allow static variables inside the function, like reading data for example, am i right?

**Comment by nzw0301** [comment]
> does Optuna run the whole objective function for each trial?

Yes, it does. 

> we shouldn't allow static variables inside the function, like reading data for example

If you'd like re-use some variables such as data samples, see https://optuna.readthedocs.io/en/stable/faq.html#how-to-define-objective-functions-that-have-own-arguments.
